transcendence-app/src/chat/chat.controller.ts

```diff
@@ -313,19 +315,17 @@ export class ChatController {
   @ApiNotFoundResponse({ description: 'Not Found' })
   @ApiServiceUnavailableResponse({ description: 'Service unavailable' })
   async getChatroomMessages(
-    @GetUser() user: User,
+    @GetAuthCrMember('chatroomId', AuthChatroomMemberPipe)
+    authUserCrm: ChatroomMemberWithAuthorization,
     @Param('chatroomId', ParseUUIDPipe) chatroomId: string,
     @Query() paginationQueryDto: PaginationQueryDto,
   ): Promise<ChatroomMessageWithUser[] | null> {
+    const ability = this.caslAbilityFactory.defineAbilitiesFor(authUserCrm);
     const chatroom = await this.chatService.getChatroomById(chatroomId);
     if (!chatroom) {
       throw new NotFoundException();
     }
-    const isChatMember = await this.chatroomMemberService.getById(
-      chatroomId,
-      user.id,
-    );
-    if (!isChatMember || isChatMember.banned) {
+    if (ability.cannot(Action.Join, chatroom)) {
```

Comment:
thought: I found this confusing because we check if the user can join a chatroom from an endpoint that lists the chatroom messages.

suggestion: What do you think about adding another `Subject` type (`'readChatroomMessagesList'`?) and using it for readability?

For example:

<details>

```diff
diff --git a/transcendence-app/src/authorization/casl-ability.factory.ts b/transcendence-app/src/authorization/casl-ability.factory.ts
index c799a6d..8625be5 100644
--- a/transcendence-app/src/authorization/casl-ability.factory.ts
+++ b/transcendence-app/src/authorization/casl-ability.factory.ts
@@ -1,7 +1,7 @@
 import {
   AbilityBuilder,
-  AnyMongoAbility,
   createMongoAbility,
+  ExtractSubjectType,
   InferSubjects,
   MongoAbility,
 } from '@casl/ability';
@@ -25,13 +25,16 @@ export type SubjectCtors =
   | typeof UserWithAuthorizationResponseDto
   | typeof ChatroomMessageWithUser;
 
-export type Subject = InferSubjects<SubjectCtors>;
+export type Subject =
+  | InferSubjects<SubjectCtors>
+  | 'all'
+  | 'readChatroomMessagesList';
 
 type AppAbility = MongoAbility<[Action, Subject]>;
 @Injectable()
 export class CaslAbilityFactory {
   private setGlobalAbilities(
-    { can, cannot, build }: AbilityBuilder<AnyMongoAbility>,
+    { can, cannot, build }: AbilityBuilder<AppAbility>,
     globalUserAuthCtx: UserWithAuthorization,
   ) {
     cannot(Action.Read, UserWithAuthorizationResponseDto);
@@ -48,7 +51,8 @@ export class CaslAbilityFactory {
       role: Role.owner,
     });
     return build({
-      detectSubjectType: (object) => object.constructor,
+      detectSubjectType: (object) =>
+        object.constructor as ExtractSubjectType<Subject>,
     });
   }
 
@@ -104,6 +108,7 @@ export class CaslAbilityFactory {
           can(Action.Read, ChatroomMessageWithUser, {
             chatroomId: user.chatId,
           });
+          can(Action.Read, 'readChatroomMessagesList');
         }
         if (user.crm_owner) {
           can(Action.Update, Chatroom, { id: user.chatId });
diff --git a/transcendence-app/src/chat/chat.controller.ts b/transcendence-app/src/chat/chat.controller.ts
index 8fffba0..1d16c7f 100644
--- a/transcendence-app/src/chat/chat.controller.ts
+++ b/transcendence-app/src/chat/chat.controller.ts
@@ -325,7 +325,7 @@ export class ChatController {
     if (!chatroom) {
       throw new NotFoundException();
     }
-    if (ability.cannot(Action.Join, chatroom)) {
+    if (ability.cannot(Action.Read, 'readChatroomMessagesList')) {
       throw new ForbiddenException();
     }
     const messages = await this.chatService.getChatroomMessagesWithUser(
diff --git a/transcendence-app/src/chat/chat.gateway.ts b/transcendence-app/src/chat/chat.gateway.ts
index b29c77b..ccbb143 100644
--- a/transcendence-app/src/chat/chat.gateway.ts
+++ b/transcendence-app/src/chat/chat.gateway.ts
@@ -125,6 +125,9 @@ export class ChatGateway {
     @ConnectedSocket() client: Socket,
   ) {
     const cr = await this.chatService.getChatroomById(chatroomId);
+    if (!cr) {
+      throw new WsException('Chatroom not found');
+    }
     const crm = await this.authorizationService.getUserAuthContextForChatroom(
       client.request.user.id,
       chatroomId,
```

</details>

Reply:
I'm unconvinced. Read is an action. The subjects should be separate and orthogonal from the actions. If you embed the action in the Subject, you defeat the purpouse of the whole system, I believe. But I think my concern is mostly the naming. I like how something like your proposal enables us to create custom abstract subjects, and how we have stricter typing now.
Anyway I implemented it the way you asked.